### ISSUE TITLE
access error when AWS_SECRET_ACCESS_KEY include "/"

### ISSUE THREAD

**Issue by fs-wu** [choplin/embulk-output-parquet#3]
It will be failed when use url format such as  <i>s3a://AWS_ACCESS_KEY_ID@AWS_SECRET_ACCESS_KEY:bucket/keys </i>
and the AWS_SECRET_ACCESS_KEY include the character "/".
Is it possible to add params such like S3 file output plugin with parameters "access_key_id","secret_access_key"?
or we wait for hadoop 2.8's launch?

**Comment by choplin** [comment]
Actually, I'm not familiar with s3 + parquet. According to the Hadoop official document, you can use configuration to set the aws keys as below. Does this work properly for parquet, too?

``` java
Configuration conf = new Configuration();
conf.set("fs.s3a.awsAccessKeyId",  "your_access_key");
conf.set("fs.s3a.awsSecretAccessKey", "your_secret_access_key");
```

If so, how about adding `extra_configuration` to enable arbitrary configuration?

``` yaml
out:
  type: parquet
  path_prefix: s3a://bucket/keys
  extra_configuration:
    - {key:fs.s3a.awsAccessKeyId,  value:your_access_key}
    - {key:fs.s3a.awsSecretAccessKey, value:your secret access key}
```

**Comment by fs-wu** [comment]
it sound good.

**Comment by choplin** [comment]
I have implemented `extra_configurations` option and pushed it to extra_configurations branch 169464b27ec1b68d6d1c95a1b76cf1bfcaa24d4d. Could you try that branch and check whether you can use it to write parquet to s3?

**Comment by fs-wu** [comment]
I had several trials but failled.
- test1  -> NG

```
path_prefix: s3a://bucket/keys
extra_configuration:
   fs.s3a.awsAccessKeyId:  key
   fs.s3a.awsSecretAccessKey: secret
```
- test2 -> NG

```
path_prefix: s3a://bucket/keys
extra_configuration:
   fs.s3a.access.key:  key
   fs.s3a.secret.key: secret
```
- test3 -> OK

```
path_prefix: s3a://key:secret@bucket/keys
```
- test4 -> NG

```
path_prefix: s3a://bucket/keys
extra_configuration:
   fs.s3.awsAccessKeyId:  key
   fs.s3.awsSecretAccessKey: secret
   fs.s3n.awsAccessKeyId:  key
   fs.s3n.awsSecretAccessKey: secret
   fs.s3a.awsAccessKeyId:  key
   fs.s3a.awsSecretAccessKey: secret
   fs.s3a.access.key:  key
   fs.s3a.secret.key: secret
```
- test5 -> NG

```
path_prefix: s3a://bucket/keys
extra_configuration:
   spark.hadoop.fs.s3a.access.key:  key
   spark.hadoop.fs.s3a.secret.key: secret
```
- test6 -> NG

```
path_prefix: s3n://bucket/keys
extra_configuration:
   fs.s3n.awsAccessKeyId:  key
   fs.s3n.awsSecretAccessKey: secret
```
- Most frequntly occured error message as:

```
org.embulk.exec.PartialExecutionException: com.amazonaws.AmazonClientException: Unable to load AWS credentials from any provider in the chain
        at org.embulk.exec.BulkLoader$LoaderState.buildPartialExecuteException(org/embulk/exec/BulkLoader.java:363)
        at org.embulk.exec.BulkLoader.doRun(org/embulk/exec/BulkLoader.java:572)
        at org.embulk.exec.BulkLoader.access$000(org/embulk/exec/BulkLoader.java:33)
        at org.embulk.exec.BulkLoader$1.run(org/embulk/exec/BulkLoader.java:374)
        at org.embulk.exec.BulkLoader$1.run(org/embulk/exec/BulkLoader.java:370)
        at org.embulk.spi.Exec.doWith(org/embulk/spi/Exec.java:25)
        at org.embulk.exec.BulkLoader.run(org/embulk/exec/BulkLoader.java:370)
        at org.embulk.EmbulkEmbed.run(org/embulk/EmbulkEmbed.java:180)
        at java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:497)
        at RUBY.run(/usr/local/bin/embulk!/embulk/runner.rb:77)
        at RUBY.run(/usr/local/bin/embulk!/embulk/command/embulk_run.rb:296)
        at RUBY.<top>(/usr/local/bin/embulk!/embulk/command/embulk_main.rb:2)
        at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:940)
        at RUBY.(root)(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:1)
        at usr.local.bin.embulk.embulk.command.embulk_bundle.<top>(file:/usr/local/bin/embulk!/embulk/command/embulk_bundle.rb:51)
        at java.lang.invoke.MethodHandle.invokeWithArguments(java/lang/invoke/MethodHandle.java:627)
        at org.embulk.cli.Main.main(org/embulk/cli/Main.java:23)
Caused by: com.amazonaws.AmazonClientException: Unable to load AWS credentials from any provider in the chain
        at com.amazonaws.auth.AWSCredentialsProviderChain.getCredentials(AWSCredentialsProviderChain.java:117)
        at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:3521)
        at com.amazonaws.services.s3.AmazonS3Client.headBucket(AmazonS3Client.java:1031)
        at com.amazonaws.services.s3.AmazonS3Client.doesBucketExist(AmazonS3Client.java:994)
        at org.apache.hadoop.fs.s3a.S3AFileSystem.initialize(S3AFileSystem.java:297)
        at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2653)
        at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:92)
        at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2687)
        at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2669)
        at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:371)
        at org.apache.hadoop.fs.Path.getFileSystem(Path.java:295)
        at parquet.hadoop.ParquetFileWriter.<init>(ParquetFileWriter.java:174)
        at parquet.hadoop.ParquetWriter.<init>(ParquetWriter.java:211)
        at parquet.hadoop.ParquetWriter.<init>(ParquetWriter.java:175)
        at org.embulk.output.ParquetOutputPlugin.createParquetWriter(ParquetOutputPlugin.java:128)
        at org.embulk.output.ParquetOutputPlugin.open(ParquetOutputPlugin.java:119)
        at org.embulk.spi.util.Executors.process(Executors.java:56)
        at org.embulk.spi.util.Executors.process(Executors.java:42)
        at org.embulk.exec.LocalExecutorPlugin$2.call(LocalExecutorPlugin.java:104)
        at org.embulk.exec.LocalExecutorPlugin$2.call(LocalExecutorPlugin.java:100)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by choplin** [comment]
Thank you for the detailed report. I am going to check how to write parquet file into s3 in my free time.

**Comment by choplin** [comment]
I found that the trials failed due to the wrong key names. The correct key names are as below, and I succeeded in writing parquet files in s3 via this plugin with the configuration as below. Could you try this?

```
  extra_configuration:
    fs.s3a.access.key: 'your_access_key'
    fs.s3a.secret.key: 'your_secret_access_key'
```

**Comment by fs-wu** [comment]
I think my test2 is the same pattern but failed.I dont know the reason, what 's your path_prefix format?

**Comment by choplin** [comment]
Here is my complete config.yml

```yaml
in:
  type: file
  path_prefix: /Users/okuno/tmp/embulk/example/csv/sample_
  decoders:
  - {type: gzip}
  parser:
    charset: UTF-8
    newline: CRLF
    type: csv
    delimiter: ','
    quote: '"'
    trim_if_not_quoted: false
    skip_header_lines: 1
    allow_extra_columns: false
    allow_optional_columns: false
    columns:
    - {name: id, type: long}
    - {name: account, type: long}
    - {name: time, type: timestamp, format: '%Y-%m-%d %H:%M:%S'}
    - {name: purchase, type: timestamp, format: '%Y%m%d'}
    - {name: comment, type: string}
out:
  type: parquet
  path_prefix: s3a://okuno-test/out_
  overwrite: true
  extra_configurations:
    fs.s3a.access.key: 'access_key'
    fs.s3a.secret.key: 'secret_key'
```

**Comment by choplin** [comment]
Sorry, I told you wrong configuration key. The correct key name is `extra_configurations`. (not `extra_configuration`)

**Comment by fs-wu** [comment]
After update the source, and update the key name, I success in my config.Please dont forget your readme for key name,thank you. Below is my config:

``` yaml
in:
  type: mysql
  host: localhost
  database: test1
  user: root
  password: root
  port: 3306
  query: select * from user
exec: {}
out:
  type: parquet
  path_prefix: s3a://temp-go/a1
  overwrite: true
  extra_configurations:
    fs.s3a.access.key: 'key'
    fs.s3a.secret.key: 'secret'
```

**Comment by choplin** [comment]
Thank you for testing. I will fix README and release it to rubygems later.

**Comment by choplin** [comment]
Released as v0.4.0
